### ISSUE TITLE
Enhancement: no-unused-vars Prevent "defined but only used as type" error

### ISSUE THREAD

**Issue by doberkofler** [typescript-eslint/typescript-eslint#9697]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://eslint.org/docs/latest/rules/no-unused-vars#options

### Description

I have a use case related to #9679 that in my code base causes the "warning  'z$foo' is assigned a value but only used as a type  @typescript-eslint/no-unused-vars" error when for example using the zod library. I get this error every time a zod definition it already used to infer the type but (not yet) to parse.
We actually generate the zod schema objects from an external definition and always export the inferred type.
The definition itself might or might not be used to parse untyped external data.
I my code base get this error a few hundred times and it would be most helpful to have an option that exclude this specific  use case of unused variable.

```typescript
const z$foo = z.object({bar: z.boolean()});
export type foo = z.infer<typeof z$foo>;
```

[playground](https://v8--typescript-eslint.netlify.app/play/#ts=5.5.2&fileType=.tsx&code=JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgNwBQFAxhAHYDO8yAJJhBHALwoB0EAIwBWAU2owAFEgEBDKAC4%2BAjgBsRMuhICU6LZREAPSLDgwAnmBFYO3PsDqYRUADznLETCjYcAfJQpAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

### Fail

```TypeScript
n.a.
```

### Pass

```TypeScript
const z$foo = z.object({bar: z.boolean()});
export type foo = z.infer<typeof z$foo>;
```

**Comment by kirkwaiblinger** [comment]
Hi @doberkofler! 

I'm interested in a bit more detail for this use case.... Why would you generate a zod schema in the first place if you don't intend to use or export it? Why not just generate `export type foo = { bar: boolean }` directly? 

And, if you do need the actual zod schema, why not just `export const z$foo = z.object({ bar: z.boolean() });`?

I guess, the overarching question I'm unsure of is whether the state you're describing is a reasonable end state for the codebase, that should be permissible by a linter _option_.... or whether it makes more sense to (temporarily?) disable entirely no-unused-vars in your config (applying to your generated files only).

So, a bit more info/discussion would be appreciated 🙏. Thanks!

**Comment by doberkofler** [comment]
Good questions!

As I mentioned the zod schema definitions are generated from an external schema and not always use for anything else than the inferred type.
We do not export them because they should only be used in the individual module in not outside but they are not always used.
Sure it would be possible to disable the warning for each definition and this is most likely what we will have to do if there is no way to disable this specific use case.
It might be an edge case but I believe that it might benefit others as well.

**Comment by kirkwaiblinger** [comment]
Ok, gotcha, thanks for clarifying 🙂 

My _personal_ feeling is that I would either disable the rule completely for these generated files (which are hopefully stored in the file system in an easily-globabble way, vaguely something like `{ files: ['src/generated-schemas/**/*.ts'], rules: { "@typescript-eslint/no-unused-vars": "off" } }`), _or_ use the unused vars ignore pattern for this case, and set it to the naming scheme for this use case (where, hopefully, the intentionally possibly-unused schemas are named with an easily-regexable naming scheme - it appears to be `^z\$.*` in your case).

So, for now, granted it applies to _generated_, not _human-authored_ code, and there are two alternatives that seem preferable to me personally, I'm still feeling -0.5 on creating this option. But other team members may feel differently... Let's see! @typescript-eslint/triage-team?

**Comment by Josh-Cena** [comment]
I'm also -1. This option would go against the spirit of the rule, which is to prevent dead code that has no runtime impacts and may have runtime overhead.

**Comment by bradzacher** [comment]
Recommendation would be to mark it as unused via a naming convention and configure the rule to ignore that.

Disabling the rule on the generated files is a good idea if you want a bigger blast radius thing.

Also one could consider using the TS api to get the final type and emit that to avoid any runtime code from being emitted. Eg "generate runtime code to a temp file, run use TS APIs to get type of inferred type, write type to codegen folder without useless runtime code." But that's like a whole next level of effort.
